src/pmu.rs

```diff
@@ -0,0 +1,301 @@
+//! PMU Extension
+#![allow(missing_docs)]
+use e310x::{BACKUP, PMU, RTC};
+
+/// Backup register size in bytes
+const BACKUP_REGISTER_BYTES: usize = 4;
+
+/// value required written to pmukey register before writing to other PMU registers
+pub const PMU_KEY_VAL: u32 = 0x51F15E;
+
+// Hifive1-revA programs
+#[cfg(not(feature = "g002"))]
+const DEFAULT_SLEEP_PROGRAM: [u32; 8] = [
+    0x0F0, // assert corerst
+    0x1F0, // assert hfclkrst
+    0x1D0, // deassert pmu_out_1
+    0x1C0, // deassert pmu_out_0
+    0x1C0, // repeats
+    0x1C0, 0x1C0, 0x1C0,
+];
+
+#[cfg(not(feature = "g002"))]
+const DEFAULT_WAKE_PROGRAM: [u32; 8] = [
+    0x1F0, // assert all resets and enable all power supplies
+    0x0F8, // idle 2^8 cycles, then deassert hfclkrst
+    0x030, // deassert corerst and padrst
+    0x030, // repeats
+    0x030, 0x030, 0x030, 0x030,
+];
+
+// Hifive1-revB programs
+#[cfg(feature = "g002")]
+const DEFAULT_SLEEP_PROGRAM: [u32; 8] = [
+    0x2F0, // assert corerst
+    0x3F0, // assert hfclkrst
+    0x3D0, // deassert pmu_out_1
+    0x3C0, // deassert pmu_out_0
+    0x3C0, // repeats
+    0x3C0, 0x3C0, 0x3C0,
+];
+
+#[cfg(feature = "g002")]
+const DEFAULT_WAKE_PROGRAM: [u32; 8] = [
+    0x3F0, // assert all resets and enable all power supplies
+    0x2F8, // idle 2^8 cycles, then deassert hfclkrst
+    0x030, // deassert corerst and padrst
+    0x030, // repeats
+    0x030, 0x030, 0x030, 0x030,
+];
+
+///
+/// Enumeration of device reset causes
+/// 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ResetCause {
+    /// Reset due to power on
+    PowerOn,
+    /// Reset due to external input (button)
+    External,
+    /// Reset due to watchdog
+    WatchDog,
+}
+
+///
+/// Enumeration of device wakeup causes
+///
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum WakeupCause {
+    /// Wake up due to reset (see ResetCause)
+    Reset(ResetCause),
+    /// Wake up due to RTC clock
+    RTC,
+    /// Wake up due to digital input (button)
+    Digital,
+}
+
+///
+/// Errors for user data backup procedures
+/// 
+#[derive(Debug)]
+pub enum BackupError {
+    /// Emitted when user data is larger than backup registers capacity
+    DataTooLarge, 
+    /// Emitted when user data size is not divisible by 4 bytes
+    DataSizeInvalid,
```

Comment:
I wonder if it's a problem at all... You can pad the data with zero bytes.

Reply:
It's a problem on restore. I had a really complex restore version that went to byte by byte for the last word (4bytes) if the alignment was off. I didn't like the possibilities.

The problem is on restore say if your struct is 6 bytes, I stored two words (8 bytes) and if I just blindly write those two words I overflow by two. Enforcing 4 which is very simple to get just makes it more clean and less error-prone.